### PR TITLE
Changes to fix bug in synthetic halo positions

### PR DESCRIPTION
**Pull request by evevkovacs**
This PR addresses issue #97 and fixes the bug in generating the positions of synthetic galaxies for healpixels that contain very few halos. Positions are no longer based on halo positions but instead on the boundaries of the healpixel.

### CODE REVIEW

File: cosmodc2/write_umachine_healpix_mock_to_disk.py [hunk 1]
```diff
@@ -352,9 +363,9 @@ def write_umachine_healpix_mock_to_disk(
             if num_infinite > 0:
                 print('...Warning: {} infinite values in synthetic {}'.format(num_infinite, m_id))
 
-        #  Now downsample the synthetic galaxies according to the
+        #  Now downsample the synthetic galaxies to adjust for volume of lightcone shell
         #  desired number = synthetic_number*comoving_vol(snapshot)/comoving_vol(healpixel)
-        volume_factor = get_volume_factor(snapshot, redshift_max, z2ts)
+        volume_factor = get_volume_factor(redshift, previous_redshift, Vtotal, cosmology)
         num_selected_synthetic = int(synthetic_number*volume_factor)
```
Comment (aphearin) [inline cosmodc2/write_umachine_healpix_mock_to_disk.py]: What happens here when volume_factor=0 ==> num_selected_synthetic=0?

Comment (evevkovacs) [reply cosmodc2/write_umachine_healpix_mock_to_disk.py]: Yes, that happens in some edge cases. Sometimes the overlap of the edge pixel with the octant is tiny (eg the healpix boundary is at x~1e-14 and we want x>0). In this case, the overlap is too small to use the MonteCarlo method to estimate the overlap. I print a message that volume_factor=0 ==> num_selected_synthetic=0 and return an empty synthetics table. This is not perfect but it is a very peculiar edge case.

File: cosmodc2/synthetic_subhalos/extend_subhalo_mpeak_range.py [hunk 2]
```diff
@@ -185,19 +186,79 @@ def create_synthetic_lowmass_mock_with_satellites(
...
+    center = np.asarray(hp.pixelfunc.pix2vec(Nside, cutout_id, nest=False)) # acccount for spherical cap
+    box_corners = np.vstack((corners.T*rmin, corners.T*rmax, center*rmax))
+    box_mins = np.asarray([np.min(box_corners[:, i]) for i in range(len(box_corners.T))])
+    box_maxs = np.asarray([np.max(box_corners[:, i]) for i in range(len(box_corners.T))])
+
+    return box_mins, box_maxs
+
+def get_volume_factor(box_mins, box_maxs, Nside, cutout_id, r_min, r_max, 
+                      volume_minx=0., volume_miny=0., volume_maxz=0., Nsample=100000):
+    volume_factor = 1.0
```
Comment (aphearin) [inline cosmodc2/synthetic_subhalos/extend_subhalo_mpeak_range.py]: is this line necessary?

File: cosmodc2/synthetic_subhalos/extend_subhalo_mpeak_range.py [hunk 3]
```diff
@@ -185,19 +186,79 @@ def create_synthetic_lowmass_mock_with_satellites(
...
+
+    corners = hp.boundaries(Nside, cutout_id, nest=False)
+    center = np.asarray(hp.pixelfunc.pix2vec(Nside, cutout_id, nest=False)) # acccount for spherical cap
+    box_corners = np.vstack((corners.T*rmin, corners.T*rmax, center*rmax))
+    box_mins = np.asarray([np.min(box_corners[:, i]) for i in range(len(box_corners.T))])
+    box_maxs = np.asarray([np.max(box_corners[:, i]) for i in range(len(box_corners.T))])
+
+    return box_mins, box_maxs
+
+def get_volume_factor(box_mins, box_maxs, Nside, cutout_id, r_min, r_max,
```
Comment (aphearin) [inline cosmodc2/synthetic_subhalos/extend_subhalo_mpeak_range.py]: There should really be a unit-test for this function.

Comment (evevkovacs) [reply cosmodc2/synthetic_subhalos/extend_subhalo_mpeak_range.py]: Yes, good catch,  I should have an additional test to check that volume_box is larger than volume_octant. If not, then the rest of the code in that function can be skipped and the volume_factor returned is 1.0

**Comment by evevkovacs** [comment]
In fact I have run tests of the sort you describe. On cooley, in /gpfs/mira-fs0/projects/DarkUniverse_esp/kovacs/OR_5000/run, you will find a logfile for running the code on healpixel 8288. The halo catalog for this healpixel is about half of the average size of a typical healpixel. The get_volume_factor function returns values between 0.498 and 0.505 for each of the snapshots processed.

**Pull request merged**